Socket leak when catching `SocketBlockedError` and running with `--allow-hosts`

With the following MRE:

```py
# test_mre.py
import urllib.request
import pytest

def test_repro() -> None:
    with pytest.raises(Exception):
        urllib.request.urlopen("https://pypi.org/project/pytest-socket")
```

Running `pytest -Werror --disable-socket test_mre.py` passes, but strangely adding any `--allow-hosts` option _(with any IP, totally unrelated to the request host)_ seem to leak a socket:

```
% uv run pytest -Werror --disable-socket --allow-hosts=1.2.3.4 test_mre.py
======================================== test session starts =========================================
platform darwin -- Python 3.14.4, pytest-9.1.1, pluggy-1.6.0
rootdir: /Users/loic/mre
configfile: pyproject.toml
plugins: socket-0.8.0
collected 1 item                                                                                     

test_mre.py F                                                                                  [100%]

============================================== FAILURES ==============================================
_____________________________________________ test_repro _____________________________________________

    def test_repro() -> None:
>       with pytest.raises(Exception):
             ^^^^^^^^^^^^^^^^^^^^^^^^
E       ResourceWarning: unclosed <socket.socket fd=14, family=2, type=1, proto=6, laddr=('0.0.0.0', 0)>

test_mre.py:5: ResourceWarning

The above exception was the direct cause of the following exception:

cls = <class '_pytest.runner.CallInfo'>
func = <function call_and_report.<locals>.<lambda> at 0x107434b40>, when = 'call'
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(
        cls,
        func: Callable[[], TResult],
        when: Literal["collect", "setup", "call", "teardown"],
        reraise: type[BaseException] | tuple[type[BaseException], ...] | None = None,
    ) -> CallInfo[TResult]:
        """Call func, wrapping the result in a CallInfo.
    
        :param func:
            The function to call. Called without arguments.
        :type func: Callable[[], _pytest.runner.TResult]
        :param when:
            The phase in which the function is called.
        :param reraise:
            Exception or exceptions that shall propagate if raised by the
            function, instead of being wrapped in the CallInfo.
        """
        excinfo = None
        instant = timing.Instant()
        try:
>           result: TResult | None = func()
                                     ^^^^^^

.venv/lib/python3.14/site-packages/_pytest/runner.py:361: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.14/site-packages/_pytest/runner.py:250: in <lambda>
    lambda: runtest_hook(item=item, **kwds),
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.14/site-packages/pluggy/_hooks.py:512: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.14/site-packages/pluggy/_manager.py:120: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.14/site-packages/_pytest/logging.py:865: in pytest_runtest_call
    yield
.venv/lib/python3.14/site-packages/_pytest/capture.py:900: in pytest_runtest_call
    return (yield)
            ^^^^^
.venv/lib/python3.14/site-packages/_pytest/skipping.py:268: in pytest_runtest_call
    return (yield)
            ^^^^^
.venv/lib/python3.14/site-packages/_pytest/unraisableexception.py:183: in pytest_runtest_call
    collect_unraisable(item.config)
.venv/lib/python3.14/site-packages/_pytest/unraisableexception.py:79: in collect_unraisable
    raise errors[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

config = <_pytest.config.Config object at 0x106a71550>

    def collect_unraisable(config: Config) -> None:
        pop_unraisable = config.stash[unraisable_exceptions].pop
        errors: list[pytest.PytestUnraisableExceptionWarning | RuntimeError] = []
        meta = None
        hook_error = None
        try:
            while True:
                try:
                    meta = pop_unraisable()
                except IndexError:
                    break
    
                if isinstance(meta, BaseException):
                    hook_error = RuntimeError("Failed to process unraisable exception")
                    hook_error.__cause__ = meta
                    errors.append(hook_error)
                    continue
    
                msg = meta.msg
                try:
>                   warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E                   pytest.PytestUnraisableExceptionWarning: Exception ignored while finalizing socket <socket.socket fd=14, family=2, type=1, proto=6, laddr=('0.0.0.0', 0)>: None

.venv/lib/python3.14/site-packages/_pytest/unraisableexception.py:67: PytestUnraisableExceptionWarning
====================================== short test summary info =======================================
FAILED test_mre.py::test_repro - pytest.PytestUnraisableExceptionWarning: Exception ignored while finalizing socket <socket.socket...
========================================= 1 failed in 0.07s ==========================================
```

Would you have an idea of why this happens?

I thought it could be due to the [`warnings.warn` emitted when constructing `UserSocketError`](https://github.com/miketheman/pytest-socket/blob/main/src/pytest_socket/__init__.py#L29), because `-Werror` would turn it into an exception, but commenting it doesn't change the outcome much.

I can try to investigate more later. 

Thanks a lot!

--- a/src/pytest_socket/__init__.py
+++ b/src/pytest_socket/__init__.py
@@ -353,6 +353,10 @@ def socket_allow_hosts(
             if any(ip in net for net in networks):
                 return _true_connect(inst, *args)
 
+        # Close the real socket before raising. The blocking error is a
+        # RuntimeError, which bypasses callers' `except OSError` cleanup
+        # (e.g. socket.create_connection), so the fd would otherwise leak.
+        inst.close()
         raise SocketConnectBlockedError(allowed_list, host)
 
     socket.socket.connect = guarded_connect  # type: ignore[assignment,method-assign]

--- a/tests/test_restrict_hosts.py
+++ b/tests/test_restrict_hosts.py
@@ -512,3 +512,26 @@ def test_name_resolution_cached(pytester, getaddrinfo_hosts):
         "name.internal": 1,
         "name.another": 1,
     }
+
+
+def test_blocked_connect_does_not_leak_socket(pytester):
+    """A blocked connect must not leak the underlying socket.
+
+    Under ``--allow-hosts`` only ``socket.socket.connect`` is guarded, so a
+    real socket is created before the connection is blocked. The block error
+    is a ``RuntimeError``, which bypasses callers' ``except OSError`` cleanup
+    (e.g. ``socket.create_connection``), leaking the file descriptor. Run with
+    ``-W error`` so the resulting ``ResourceWarning`` fails the test.
+    """
+    pytester.makepyfile("""
+        import pytest
+        from urllib.request import urlopen
+
+        def test_blocked():
+            with pytest.raises(Exception):
+                # IP literal avoids DNS; a real socket is created then blocked.
+                urlopen("http://1.2.3.5/")
+    """)
+
+    result = pytester.runpytest("-W", "error", "--allow-hosts=1.2.3.4")
+    result.assert_outcomes(passed=1)

